lib/src/entrypoint.dart

```diff
@@ -642,4 +646,127 @@ class Entrypoint {
+    var localScriptCache = p.equals(cache.rootDir, cachePath)
+        ? systemScriptCache
+        : await AfterInstallCache.load(cachePath);
+    var mergedCache = systemScriptCache.merge(localScriptCache);
+
+    // Sort so that the root's scripts run last.
+    var changedPackages = result.changedPackages.toList();
+    changedPackages.sort((a, b) => a == root.name ? 1 : a.compareTo(b));
+
+    // Run scripts in dependencies.
+    for (var package in changedPackages) {
+      var pkg = package == root.name
+          ? root
+          : cache.load(result.packages
+              .firstWhere((a) => a.name == package, orElse: () => null));
+
+      for (var script in pkg.pubspec.afterInstall) {
+        var absolutePath = p.normalize(p.absolute(p.join(pkg.dir, script)));
+        var scriptFile = new File(absolutePath);
+
+        // Don't run the script against there are new changes.
+        if (!await mergedCache.isOutdated(absolutePath)) continue;
+
+        if (!await scriptFile.exists()) {
+          if (pkg == root)
+            log.warning(log.yellow(
+                '"after_install" script "$absolutePath" does not exist.'));
+        } else {
+          // Never trust any third-party scripts. Show a prompt before running scripts.
+          //
+          // The user can opt to print the contents of the script.
+          YesNoPrintResponse response = YesNoPrintResponse.yes;
+
+          if (pkg != root) {
+            var message =
+                '\npackage:$package wants to run the script "$absolutePath".';
+
+            do {
+              response = await confirmOrPrint(message);
+
+              if (response == YesNoPrintResponse.no) {
+                log.message('User declined to run script.');
+                continue;
+              }
+
+              if (response == YesNoPrintResponse.print) {
+                log.message(log.gray('Contents of "$absolutePath":'));
+                log.message(log.gray(await scriptFile.readAsString()));
+              }
+            } while (response == YesNoPrintResponse.print);
+          }
+
+          if (response != YesNoPrintResponse.yes) continue;
+
+          // We need to change into the package's directory to run its scripts.
+          var currentDir = Directory.current;
+          Directory.current = pkg.dir;
+
+          // Spawn an isolate for the script.
+          var onError = new ReceivePort();
+          var onExit = new ReceivePort();
+          var onComplete = new Completer();
+
+          onExit.listen(
+              (x) => onComplete.isCompleted ? null : onComplete.complete());
+          onError.listen((x) => onComplete.isCompleted
+              ? null
+              : onComplete.completeError(x[0], x[1]));
+
+          try {
+            Isolate.spawnUri(p.toUri(absolutePath), [], null,
+                onExit: onExit.sendPort,
+                onError: onError.sendPort,
+                automaticPackageResolution: true);
```

Comment:
We should use the `packageConfig` argument and pass in the config from the current isolate, instead of `automaticPackageResolution` argument, similar to `pub run` behavior. This likely gets around the issue that was causing you to change directories to the other package?

Reply:
That's probably the solution I was looking for, haha. Is there a way to specify a working directory for an isolate, though? In my head, the scripts would execute in the folder for the specific version that package, so that things like native extensions could be built to the correct paths.

Otherwise, though, people publishing scripts could still just `Isolate.resolvePackageUri`, probably

Reply:
> In my head, the scripts would execute in the folder for the specific version that package, so that things like native extensions could be built to the correct paths.

We really don't want to give access to the actual package cache. That should only contain the exact sources as they exist on pub because it is shared between all projects on your machine. The output of the `after_install` script could be different depending on the version solve of any given project. Therefore each individual project should have to re-run any `after_install` scripts and should generally write the results to the `.dart_tool/<package-name>` directory (by convention).

Reply:
Okay, that's fair, and makes a lot of sense.

One more question: what can end-users do to depend on sources built to `.dart_tool`? By passing a custom package config? I'm not 100% sure how `build` does it.

Reply:
That is actually a really good question haha - basically the build system knows how to surface those files via the `serve` command and within the build and treat it as an overlay file system. That isn't a general solution for all of .dart_tool though, its very specific to the `.dart_tool/build/generated` directory and the exact files that it knows it wrote there.

We would have to come up with some other solution here for pub, possibly writing custom paths to the `.packages` file that point to `.dart_tool` and copying the original package + generated sources there.